src/control/drpc/drpc_server.go

```diff
@@ -0,0 +1,202 @@
+			d.closeSession(s)
+			break
+		}
+	}
+}
+
+// Listen listens for incoming connections on the UNIX domain socket and
+// creates individual sessions for each one.
+func (d *DomainSocketServer) Listen() {
+	for {
+		conn, err := d.listener.Accept()
+		if err != nil {
+			select {
+			case <-d.quit:
+				d.log.Debug("Quitting listener")
+			default:
+				d.log.Errorf("%s: failed to accept connection: %v", d.sockFile, err)
+			}
+
+			return
+		}
+
+		d.log.Debug("Creating session for connection")
+		c := NewSession(conn, d.service)
+		d.sessions[conn] = c
+		go d.listenSession(c)
+	}
+}
+
+// Start sets up the dRPC server socket and kicks off the listener goroutine.
+func (d *DomainSocketServer) Start() error {
+	// Just in case an old socket file is still lying around
+	if err := syscall.Unlink(d.sockFile); err != nil && !os.IsNotExist(err) {
+		return errors.Wrapf(err, "Unable to unlink %s", d.sockFile)
+	}
+
+	addr := &net.UnixAddr{Name: d.sockFile, Net: "unixpacket"}
+	lis, err := net.ListenUnix("unixpacket", addr)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to listen on unix socket %s", d.sockFile)
+	}
+	d.listener = lis
+
+	// TODO: Should we set more granular permissions? The only writer should
+	// be the IO server and we should know which user is running it.
+	if err := os.Chmod(d.sockFile, 0777); err != nil {
+		return errors.Wrapf(err, "Unable to set permissions on %s", d.sockFile)
+	}
+
+	go d.Listen()
+	return nil
+}
+
+// Shutdown places the state of the server to shutdown which terminates the
+// Listen go routine and starts the cleanup of all open connections.
+func (d *DomainSocketServer) Shutdown() {
+	close(d.quit)
+
+	for _, s := range d.sessions {
+		d.closeSession(s)
+	}
+	d.listener.Close()
+}
+
+// RegisterRPCModule takes a Module and associates it with the given
+// DomainSocketServer so it can be used to process incoming dRPC calls.
+func (d *DomainSocketServer) RegisterRPCModule(mod Module) {
+	d.service.RegisterModule(mod)
+}
+
+// NewDomainSocketServer returns a new unstarted instance of a
+// DomainSocketServer for the specified unix domain socket path.
+func NewDomainSocketServer(log logging.Logger, sock string) (*DomainSocketServer, error) {
+	if sock == "" {
+		return nil, errors.New("Missing Argument: sockFile")
+	}
+	service := NewModuleService(log)
+	quit := make(chan bool)
+	sessions := make(map[net.Conn]*Session)
+	return &DomainSocketServer{
+		log:      log,
+		sockFile: sock,
+		quit:     quit,
+		service:  service,
+		sessions: sessions}, nil
+}
+
+// Session represents an individual client connection to the Domain Socket Server.
+type Session struct {
+	Conn net.Conn
+	mod  *ModuleService
+}
+
+// ProcessIncomingMessage listens for an incoming message on the session,
+// calls its handler, and sends the response.
+func (s *Session) ProcessIncomingMessage() error {
+	buffer := make([]byte, MAXMSGSIZE)
+
+	bytesRead, err := s.Conn.Read(buffer)
```

Comment:
This should also not be the same semantically. Read/Write I don't believe behave the same way as ReadMsg/WriteMsg